File: ios/Runner/MetaAppEventsBridge.swift
import Flutter
import UIKit
import FBSDKCoreKit
import AppTrackingTransparency

/// MethodChannel bridge for the official Meta/Facebook App Events iOS SDK.
enum MetaAppEventsBridge {
  static let channelName = "com.app.deenly.deenly/meta_app_events"

  private static var initialized = false

  static func register(with messenger: FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { call, result in
      switch call.method {
      case "initialize":
        let args = call.arguments as? [String: Any] ?? [:]
        let ok = initialize(
          appId: (args["appId"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "",
          clientToken: (args["clientToken"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "",
          displayName: (args["displayName"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "Deen Focus",
          debug: args["debug"] as? Bool ?? false
        )
        result(["ok": ok, "initialized": initialized])
      case "isInitialized":
        result(initialized)
      case "trackAppLaunch":
        trackAppLaunch()
        result(true)
      case "trackPurchase":
        trackPurchase(args: call.arguments as? [String: Any] ?? [:])
        result(true)
      case "trackSubscribe":
        trackStandard(name: .subscribe, args: call.arguments as? [String: Any] ?? [:])
        result(true)
      case "trackStartTrial":
        trackStandard(name: .startTrial, args: call.arguments as? [String: Any] ?? [:])
        result(true)
      case "setAdvertiserTrackingEnabled":
        let enabled = (call.arguments as? [String: Any])?["enabled"] as? Bool ?? false
        setAdvertiserTrackingEnabled(enabled)
        result(true)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  /// Call from AppDelegate so deep links / URL opens are attributed.
  static func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) {
    ApplicationDelegate.shared.application(
      application,
      didFinishLaunchingWithOptions: launchOptions
    )
  }

  @discardableResult
  static func application(
    _ app: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey: Any] = [:]
  ) -> Bool {
    ApplicationDelegate.shared.application(
      app,
      open: url,
      sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String,
      annotation: options[UIApplication.OpenURLOptionsKey.annotation]
    )
  }

  @discardableResult
  private static func initialize(
    appId: String,
    clientToken: String,
    displayName: String,
    debug: Bool
  ) -> Bool {
    if initialized { return true }
    guard !appId.isEmpty, !clientToken.isEmpty else {
      NSLog("[MetaAppEvents] Skipping init; missing appId/clientToken")
      return false
    }

    Settings.shared.appID = appId
    Settings.shared.clientToken = clientToken
    Settings.shared.displayName = displayName.isEmpty ? "Deen Focus" : displayName
    Settings.shared.isAutoLogAppEventsEnabled = true
    Settings.shared.isAdvertiserIDCollectionEnabled = true
    if debug {
      Settings.shared.enableLoggingBehavior(.appEvents)
    }

    requestTrackingAuthorizationIfNeeded()
    AppEvents.shared.activateApp()
    initialized = true
    NSLog("[MetaAppEvents] Meta SDK initialized")
    return true
  }

  private static func requestTrackingAuthorizationIfNeeded() {
    guard #available(iOS 14, *) else {
      Settings.shared.isAdvertiserTrackingEnabled = true
      return
    }

    DispatchQueue.main.async {
      ATTrackingManager.requestTrackingAuthorization { status in
        let enabled = status == .authorized
        Settings.shared.isAdvertiserTrackingEnabled = enabled
        NSLog(
          "[MetaAppEvents] ATT status=%d advertiserTracking=%@",
          status.rawValue,
          enabled ? "true" : "false"
        )
      }
    }
  }

  private static func setAdvertiserTrackingEnabled(_ enabled: Bool) {
    Settings.shared.isAdvertiserTrackingEnabled = enabled
  }

  private static func trackAppLaunch() {
    guard initialized else { return }
    AppEvents.shared.activateApp()
  }

  private static func trackPurchase(args: [String: Any]) {
    guard initialized else { return }
    guard let value = number(args["value"]) else { return }
    guard let currency = string(args["currency"]), !currency.isEmpty else { return }

    var params: [AppEvents.ParameterName: Any] = [:]
    if let contentId = string(args["contentId"]), !contentId.isEmpty {
      params[.contentID] = contentId
    }
    params[.contentType] = string(args["contentType"]).flatMap { $0.isEmpty ? nil : $0 } ?? "product"
    if let eventId = string(args["eventId"]), !eventId.isEmpty {
      params[AppEvents.ParameterName("fb_order_id")] = eventId
    }

    AppEvents.shared.logPurchase(
      amount: value,
      currency: currency.uppercased(),
      parameters: params
    )
  }

  private static func trackStandard(name: AppEvents.Name, args: [String: Any]) {
    guard initialized else { return }

    var params: [AppEvents.ParameterName: Any] = [:]
    if let contentId = string(args["contentId"]), !contentId.isEmpty {
      params[.contentID] = contentId
    }
    if let currency = string(args["currency"]), !currency.isEmpty {
      params[.currency] = currency.uppercased()
    }
    if let eventId = string(args["eventId"]), !eventId.isEmpty {
      params[AppEvents.ParameterName("fb_order_id")] = eventId
    }

    if let value = number(args["value"]) {
      AppEvents.shared.logEvent(name, valueToSum: value, parameters: params)
    } else {
      AppEvents.shared.logEvent(name, parameters: params)
    }
  }

  private static func string(_ value: Any?) -> String? {
    if let s = value as? String {
      return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    return nil
  }

  private static func number(_ value: Any?) -> Double? {
    if let n = value as? NSNumber { return n.doubleValue }
    if let d = value as? Double { return d }
    if let i = value as? Int { return Double(i) }
    if let s = value as? String { return Double(s) }
    return nil
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/MetaAppEventsBridge.kt
package com.rnr.deenfocus

import android.app.Application
import android.content.Context
import android.os.Bundle
import android.util.Log
import com.facebook.FacebookSdk
import com.facebook.LoggingBehavior
import com.facebook.appevents.AppEventsConstants
import com.facebook.appevents.AppEventsLogger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.math.BigDecimal
import java.util.Currency

/**
 * MethodChannel bridge for the official Meta/Facebook App Events Android SDK.
 * Failures are swallowed so tracking never crashes the app.
 */
object MetaAppEventsBridge : MethodChannel.MethodCallHandler {
    const val CHANNEL = "com.app.deenly.deenly/meta_app_events"

    private const val TAG = "MetaAppEvents"

    @Volatile
    private var initialized = false

    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var logger: AppEventsLogger? = null

    fun attach(context: Context) {
        appContext = context.applicationContext
    }

    /** Early Application init from BuildConfig when credentials are present. */
    fun initializeFromBuildConfig(application: Application) {
        val appId = BuildConfig.META_APP_ID.trim()
        val clientToken = BuildConfig.META_CLIENT_TOKEN.trim()
        if (appId.isEmpty() || clientToken.isEmpty()) {
            Log.i(TAG, "Skipping early Meta init; credentials not configured")
            return
        }
        initializeInternal(
            application = application,
            appId = appId,
            clientToken = clientToken,
            displayName = BuildConfig.META_DISPLAY_NAME.trim().ifEmpty { "Deen Focus" },
            debug = BuildConfig.DEBUG,
        )
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "initialize" -> {
                    val context = appContext
                    if (context == null) {
                        result.success(mapOf("ok" to false, "reason" to "no_context"))
                        return
                    }
                    val ok = initializeInternal(
                        application = context.applicationContext as Application,
                        appId = call.argument<String>("appId").orEmpty().trim(),
                        clientToken = call.argument<String>("clientToken").orEmpty().trim(),
                        displayName = call.argument<String>("displayName").orEmpty().trim()
                            .ifEmpty { "Deen Focus" },
                        debug = call.argument<Boolean>("debug") ?: false,
                    )
                    result.success(mapOf("ok" to ok, "initialized" to initialized))
                }
                "isInitialized" -> result.success(initialized)
                "trackAppLaunch" -> {
                    trackAppLaunch()
                    result.success(true)
                }
                "trackPurchase" -> {
                    trackPurchase(call)
                    result.success(true)
                }
                "trackSubscribe" -> {
                    trackStandard(
                        eventName = AppEventsConstants.EVENT_NAME_SUBSCRIBE,
                        call = call,
                    )
                    result.success(true)
                }
                "trackStartTrial" -> {
                    trackStandard(
                        eventName = AppEventsConstants.EVENT_NAME_START_TRIAL,
                        call = call,
                    )
                    result.success(true)
                }
                "setAdvertiserTrackingEnabled" -> {
                    // Android uses AdvertiserIDCollectionEnabled; no ATT equivalent.
                    result.success(true)
                }
                else -> result.notImplemented()
            }
        } catch (t: Throwable) {
            Log.w(TAG, "MethodChannel ${call.method} failed safely", t)
            result.success(false)
        }
    }

    private fun initializeInternal(
        application: Application,
        appId: String,
        clientToken: String,
        displayName: String,
        debug: Boolean,
    ): Boolean {
        if (initialized) return true
        if (appId.isEmpty() || clientToken.isEmpty()) {
            Log.i(TAG, "Initialize skipped; missing META_APP_ID / META_CLIENT_TOKEN")
            return false
        }

        return try {
            FacebookSdk.setApplicationId(appId)
            FacebookSdk.setClientToken(clientToken)
            FacebookSdk.setApplicationName(displayName)
            // Keep Install/Launch auto-logging; conversion events are manual after Superwall.
            FacebookSdk.setAutoInitEnabled(true)
            FacebookSdk.setAutoLogAppEventsEnabled(true)
            FacebookSdk.setAdvertiserIDCollectionEnabled(true)
            FacebookSdk.sdkInitialize(application)
            FacebookSdk.fullyInitialize()

            if (debug) {
                FacebookSdk.setIsDebugEnabled(true)
                FacebookSdk.addLoggingBehavior(LoggingBehavior.APP_EVENTS)
            }

            logger = AppEventsLogger.newLogger(application)
            // ActivateApp / launch is also auto-logged; this reinforces session start.
            AppEventsLogger.activateApp(application)
            initialized = true
            Log.i(TAG, "Meta SDK initialized")
            true
        } catch (t: Throwable) {
            Log.w(TAG, "Meta SDK initialize threw; continuing without tracking", t)
            false
        }
    }

    private fun trackAppLaunch() {
        if (!initialized) return
        try {
            val application = appContext?.applicationContext as? Application ?: return
            AppEventsLogger.activateApp(application)
        } catch (t: Throwable) {
            Log.w(TAG, "trackAppLaunch failed safely", t)
        }
    }

    private fun trackPurchase(call: MethodCall) {
        if (!initialized) return
        try {
            val value = (call.argument<Number>("value") ?: return).toDouble()
            val currencyCode = call.argument<String>("currency").orEmpty().trim()
            if (currencyCode.isEmpty()) return

            val params = Bundle()
            val contentId = call.argument<String>("contentId").orEmpty().trim()
            val contentType = call.argument<String>("contentType").orEmpty().trim()
                .ifEmpty { "product" }
            val eventId = call.argument<String>("eventId").orEmpty().trim()
            if (contentId.isNotEmpty()) {
                params.putString(AppEventsConstants.EVENT_PARAM_CONTENT_ID, contentId)
            }
            params.putString(AppEventsConstants.EVENT_PARAM_CONTENT_TYPE, contentType)
            if (eventId.isNotEmpty()) {
                params.putString(AppEventsConstants.EVENT_PARAM_ORDER_ID, eventId)
            }

            logger?.logPurchase(
                BigDecimal.valueOf(value),
                Currency.getInstance(currencyCode.uppercase()),
                params,
            )
        } catch (t: Throwable) {
            Log.w(TAG, "trackPurchase failed safely", t)
        }
    }

    private fun trackStandard(eventName: String, call: MethodCall) {
        if (!initialized) return
        try {
            val params = Bundle()
            val contentId = call.argument<String>("contentId").orEmpty().trim()
            val currencyCode = call.argument<String>("currency").orEmpty().trim()
            val eventId = call.argument<String>("eventId").orEmpty().trim()
            val value = call.argument<Number>("value")?.toDouble()

            if (contentId.isNotEmpty()) {
                params.putString(AppEventsConstants.EVENT_PARAM_CONTENT_ID, contentId)
            }
            if (currencyCode.isNotEmpty()) {
                params.putString(AppEventsConstants.EVENT_PARAM_CURRENCY, currencyCode.uppercase())
            }
            if (eventId.isNotEmpty()) {
                params.putString(AppEventsConstants.EVENT_PARAM_ORDER_ID, eventId)
            }

            if (value != null) {
                logger?.logEvent(eventName, value, params)
            } else {
                logger?.logEvent(eventName, params)
            }
        } catch (t: Throwable) {
            Log.w(TAG, "trackStandard $eventName failed safely", t)
        }
    }
}
